File: 2016/day11.js
/*

--- Day 11: Radioisotope Thermoelectric Generators ---

You come upon a column of four floors that have been entirely sealed off from the rest of the building except for a small dedicated lobby. There are some radiation warnings and a big sign which reads "Radioisotope Testing Facility".

According to the project status board, this facility is currently being used to experiment with Radioisotope Thermoelectric Generators (RTGs, or simply "generators") that are designed to be paired with specially-constructed microchips. Basically, an RTG is a highly radioactive rock that generates electricity through heat.

The experimental RTGs have poor radiation containment, so they're dangerously radioactive. The chips are prototypes and don't have normal radiation shielding, but they do have the ability to generate an electromagnetic radiation shield when powered. Unfortunately, they can only be powered by their corresponding RTG. An RTG powering a microchip is still dangerous to other microchips.

In other words, if a chip is ever left in the same area as another RTG, and it's not connected to its own RTG, the chip will be fried. Therefore, it is assumed that you will follow procedure and keep chips connected to their corresponding RTG when they're in the same room, and away from other RTGs otherwise.

These microchips sound very interesting and useful to your current activities, and you'd like to try to retrieve them. The fourth floor of the facility has an assembling machine which can make a self-contained, shielded computer for you to take with you - that is, if you can bring it all of the RTGs and microchips.

Within the radiation-shielded part of the facility (in which it's safe to have these pre-assembly RTGs), there is an elevator that can move between the four floors. Its capacity rating means it can carry at most yourself and two RTGs or microchips in any combination. (They're rigged to some heavy diagnostic equipment - the assembling machine will detach it for you.) As a security measure, the elevator will only function if it contains at least one RTG or microchip. The elevator always stops on each floor to recharge, and this takes long enough that the items within it and the items on that floor can irradiate each other. (You can prevent this if a Microchip and its Generator end up on the same floor in this way, as they can be connected while the elevator is recharging.)

You make some notes of the locations of each component of interest (your puzzle input). Before you don a hazmat suit and start moving things around, you'd like to have an idea of what you need to do.

When you enter the containment area, you and the elevator will start on the first floor.

For example, suppose the isolated area has the following arrangement:

The first floor contains a hydrogen-compatible microchip and a lithium-compatible microchip.
The second floor contains a hydrogen generator.
The third floor contains a lithium generator.
The fourth floor contains nothing relevant.
As a diagram (F# for a Floor number, E for Elevator, H for Hydrogen, L for Lithium, M for Microchip, and G for Generator), the initial state looks like this:

F4 .  .  .  .  .  
F3 .  .  .  LG .  
F2 .  HG .  .  .  
F1 E  .  HM .  LM 

Then, to get everything up to the assembling machine on the fourth floor, the following steps could be taken:

Bring the Hydrogen-compatible Microchip to the second floor, which is safe because it can get power from the Hydrogen Generator:

F4 .  .  .  .  .  
F3 .  .  .  LG .  
F2 E  HG HM .  .  
F1 .  .  .  .  LM 

Bring both Hydrogen-related items to the third floor, which is safe because the Hydrogen-compatible microchip is getting power from its generator:

F4 .  .  .  .  .  
F3 E  HG HM LG .  
F2 .  .  .  .  .  
F1 .  .  .  .  LM 

Leave the Hydrogen Generator on floor three, but bring the Hydrogen-compatible Microchip back down with you so you can still use the elevator:

F4 .  .  .  .  .  
F3 .  HG .  LG .  
F2 E  .  HM .  .  
F1 .  .  .  .  LM 

At the first floor, grab the Lithium-compatible Microchip, which is safe because Microchips don't affect each other:

F4 .  .  .  .  .  
F3 .  HG .  LG .  
F2 .  .  .  .  .  
F1 E  .  HM .  LM 

Bring both Microchips up one floor, where there is nothing to fry them:

F4 .  .  .  .  .  
F3 .  HG .  LG .  
F2 E  .  HM .  LM 
F1 .  .  .  .  .  

Bring both Microchips up again to floor three, where they can be temporarily connected to their corresponding generators while the elevator recharges, preventing either of them from being fried:

F4 .  .  .  .  .  
F3 E  HG HM LG LM 
F2 .  .  .  .  .  
F1 .  .  .  .  .  

Bring both Microchips to the fourth floor:

F4 E  .  HM .  LM 
F3 .  HG .  LG .  
F2 .  .  .  .  .  
F1 .  .  .  .  .  

Leave the Lithium-compatible microchip on the fourth floor, but bring the Hydrogen-compatible one so you can still use the elevator; this is safe because although the Lithium Generator is on the destination floor, you can connect Hydrogen-compatible microchip to the Hydrogen Generator there:

F4 .  .  .  .  LM 
F3 E  HG HM LG .  
F2 .  .  .  .  .  
F1 .  .  .  .  .  

Bring both Generators up to the fourth floor, which is safe because you can connect the Lithium-compatible Microchip to the Lithium Generator upon arrival:

F4 E  HG .  LG LM 
F3 .  .  HM .  .  
F2 .  .  .  .  .  
F1 .  .  .  .  .  

Bring the Lithium Microchip with you to the third floor so you can use the elevator:

F4 .  HG .  LG .  
F3 E  .  HM .  LM 
F2 .  .  .  .  .  
F1 .  .  .  .  .  

Bring both Microchips to the fourth floor:

F4 E  HG HM LG LM 
F3 .  .  .  .  .  
F2 .  .  .  .  .  
F1 .  .  .  .  .  

In this arrangement, it takes 11 steps to collect all of the objects at the fourth floor for assembly. (Each elevator stop counts as one step, even if nothing is added to or removed from it.)

In your situation, what is the minimum number of steps required to bring all of the objects to the fourth floor?


--- Part Two ---

You step into the cleanroom separating the lobby from the isolated area and put on the hazmat suit.

Upon entering the isolated containment area, however, you notice some extra parts on the first floor that weren't listed on the record outside:

An elerium generator.
An elerium-compatible microchip.
A dilithium generator.
A dilithium-compatible microchip.

These work just like the other generators and microchips. You'll have to get them up to assembly as well.

What is the minimum number of steps required to bring all of the objects, including these four new ones, to the fourth floor?

*/

const { Queue } = require('./_classes');

// OPTIONAL VARIABLES
const DISPLAY_EXTRA_INFO = true;

function moveStuffToTopFloorWithBFS (part, inputStr, DEBUG = false) {
  const inputArr = inputStr.split('\r\n');

  // CONSTANTS
  const REF = {
    'H': 'H',
    'L': 'L',
    'polonium': 'Po',
    'thulium': 'Tm',
    'promethium': 'Pm',
    'ruthenium': 'Ru',
    'cobalt': 'Co',
    'elerium': 'El',
    'dilithium': 'Dl',
  };

  const ELEVATOR = 'E';

  // UTILITY FUNCTION: GIVEN A STATE ARRAY, SORT ALL ITEMS WITHIN EACH FLOOR, AND STRINGIFY SO THAT IT CAN BE COMPARED WITH VISITED SET AND WINNING STATE
  function serialize(state) {
    return state.map(floor => floor.sort()).join('|');
  }

  // UTILITY FUNCTION: GIVEN A FLOOR CONFIGURATION, MAKE SURE THERE ARE NO UNPAIRED MICROCHIPS IN THE PRESENCE OF INCOMPATIBLE GENERATORS
  function verifyFloor(floor) {
    const microchipsOnFloor = floor.filter(item => item.at(-1) === 'M');
    const generatorsOnFloor = floor.filter(item => item.at(-1) === 'G');
    for (const microchip of microchipsOnFloor) {
      const type = microchip.slice(0, microchip.length - 1);
      const generator = type + 'G';
      if (generatorsOnFloor.length && !floor.includes(generator)) return false;                                   // FALSE if there ARE generators, but not the compatible one for the current microchip
    }
    return true;                                                                                                  // TRUE if no problems with any microchips
  }

  // PARSE INPUT DATA, GENERATE INITIAL STATE
  const initialState = [];
  const generators = [];                                                                                          // track all generators to create winning state
  const microchips = [];                                                                                          // track all microchips to create winning state
  for (const line of inputArr) {
    const floor = [];
    const split = line.split(' ');
    for (let i = 0; i < split.length; ++i) {
      if (split[i].slice(0, 9) === 'generator') {                                                                 // NOTE: slice required to handle period at end of sentence
        const generator = REF[split[i - 1]] + 'G';
        generators.push(generator);
        floor.push(generator);
      }
      if (split[i].slice(0, 9) === 'microchip') {
        const microchip = REF[split[i - 1].split('-')[0]] + 'M';
        microchips.push(microchip);
        floor.push(microchip);
      }
    }
    initialState.push(floor);                                                                                     // NOTE: assume input data provides information about the floors in order from first to fourth
  }
  if (part === 2) {                                                                                               // PART 2: ADD 4 NEW ITEMS
    const eleriumGenerator = REF['elerium'] + 'G';
    const eleriumMicrochip = REF['elerium'] + 'M';
    const dilithiumGenerator = REF['dilithium'] + 'G';
    const dilithiumMicrochip = REF['dilithium'] + 'M';

    generators.push(eleriumGenerator);
    generators.push(dilithiumGenerator);
    microchips.push(eleriumMicrochip);
    microchips.push(dilithiumMicrochip);

    initialState[0].push(eleriumGenerator);
    initialState[0].push(eleriumMicrochip);
    initialState[0].push(dilithiumGenerator);
    initialState[0].push(dilithiumMicrochip);
  }
  initialState[0].push(ELEVATOR);                                                                                 // NOTE: don't forget to include elevator

  const winState = serialize([ [], [], [], [ ELEVATOR, ...generators, ...microchips ] ]);                         // win state has everything on 4th floor

  // INIT BFS
  const Q = new Queue([ initialState, 0 ]);
  const visited = new Set();

  // ANALYZE BFS
  const TIME_AT_START = Date.now();
  if (!DEBUG) console.log('RUNNING ANALYSIS (PLEASE WAIT)...');
  let NEXT_MIN_TARGET = 1;
  while (!Q.isEmpty()) {

    if (DISPLAY_EXTRA_INFO
      && Math.floor((Date.now() - TIME_AT_START)/(1000*60)) === NEXT_MIN_TARGET)
    {
      console.log(`... ${
        Math.floor((Date.now() - TIME_AT_START)/(1000*60))
      } mins have passed since beginning this run`);
      ++NEXT_MIN_TARGET;
    }

    // dequeue and serialize state
    const [ state, moves ] = Q.dequeue().val;
    const serial = serialize(state);

    // skip visited states
    if (visited.has(serial)) continue;
    visited.add(serial);

    // check if win condition
    if (serial === winState) {
      if (!DEBUG) console.log(`(RUN TOOK ${(Date.now() - TIME_AT_START)/1000} SECS)`);
      return moves;
    }

    // find all possible neighboring states from here
    let elevatorFloor;
    for (let i = 0; i < 4; ++i) {
      if (state[i].includes(ELEVATOR)) {
        elevatorFloor = i;
        break;
      }
    }
    const neighboringFloors = [elevatorFloor - 1, elevatorFloor + 1].filter(floor => 0 <= floor && floor < 4);    // check above and below (if floor in bounds)

    const itemsOnThisFloor = state[elevatorFloor].filter(item => item !== ELEVATOR);                              // do not include elevator for movable items
    
    // case A: take one thing only
    for (const takenItem of itemsOnThisFloor) {

      const takenItems = [ ELEVATOR, takenItem ];                                                                 // NOTE: add back the elevator itself as an 'item', since it moves, for purposes of the new state

      if (verifyFloor(itemsOnThisFloor.filter(item => !takenItems.includes(item)))) {                             // first, verify no issues with this floor if items taken away
        for (const neighboringFloor of neighboringFloors) {
          if (verifyFloor([ ...state[neighboringFloor], ...takenItems ])) {                                       // then, verify no issues if items brought to candidate floor
            Q.enqueue([                                                                                           // if all checks work out, enqueue this possibility
              state.map((_, floor) => {                                                                           // NOTE: must DEEP CLONE the state
                if (floor === elevatorFloor) return state[floor].filter(item => !takenItems.includes(item));      // remove taken items from current floor
                if (floor === neighboringFloor) return [ ...state[floor], ...takenItems ];                        // add taken items to destination floor
                return [ ...state[floor] ];                                                                       // leave all other floors the same
              }),
              moves + 1,                                                                                          // increment moves
            ]);
          }
        }
      }
    }

    // case B: take two things
    for (let i = 0; i < itemsOnThisFloor.length - 1; ++i) {
      for (let j = i + 1; j < itemsOnThisFloor.length; ++j) {

        const takenItems = [ ELEVATOR, itemsOnThisFloor[i], itemsOnThisFloor[j] ];

        if (verifyFloor(itemsOnThisFloor.filter(item => !takenItems.includes(item)))) {
          for (const neighboringFloor of neighboringFloors) {
            if (verifyFloor([ ...state[neighboringFloor], ...takenItems ])) {
              Q.enqueue([
                state.map((_, floor) => {
                  if (floor === elevatorFloor) return state[floor].filter(item => !takenItems.includes(item));
                  if (floor === neighboringFloor) return [ ...state[floor], ...takenItems ];
                  return [ ...state[floor] ];
                }),
                moves + 1,
              ]);
            }
          }
        }
      }
    }
  }

  if (!DEBUG) console.log(`(RUN TOOK ${(Date.now() - TIME_AT_START)/1000} SECS)`);
  throw 'ERROR: DID NOT FIND SOLUTION';
}

// TEST CASES

const test = require('./_test');
const testNum = [1];
let input, expected;
const func = moveStuffToTopFloorWithBFS;
const sortedFunc = (...args) => func(...args).sort();                   // used when the order of the output does not matter
const modFunc = (...args) => func(...args) % 1000000007;                // used when the output is very large
const skippedTests = new Set([  ]);
const lowestTest = 0;
const highestTest = 0;

const fs = require('fs');
const path = require('path');
const DAY_NUM = __filename.split('.js')[0].split('day')[1];
const INPUT_PATH = path.join(__dirname, `day${DAY_NUM}-input.txt`);
const actualInput = fs.readFileSync(INPUT_PATH, 'utf8');
const parseSampleInput = s => s.split('').map(c => c === '\n' ? '\r\n' : c).join('');

const sampleInput = parseSampleInput(
`The first floor contains a H-compatible microchip and a L-compatible microchip.
The second floor contains a H generator.
The third floor contains a L generator.
The fourth floor contains nothing relevant.`
);

// Test case 1
input = {
  part: 1,
  inputStr: sampleInput,
  DEBUG: true,
};
expected = 11;
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

// Test case 2
input = {
  part: 1,
  inputStr: actualInput,
};
expected = 47;
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

// Test case 3
input = {
  part: 2,
  inputStr: actualInput,
};
expected = 71;
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

File: 2023/day11.py
"""

--- Day 11: Cosmic Expansion ---

You continue following signs for "Hot Springs" and eventually come across an observatory. The Elf within turns out to be a researcher studying cosmic expansion using the giant telescope here.

He doesn't know anything about the missing machine parts; he's only visiting for this research project. However, he confirms that the hot springs are the next-closest area likely to have people; he'll even take you straight there once he's done with today's observation analysis.

Maybe you can help him with the analysis to speed things up?

The researcher has collected a bunch of data and compiled the data into a single giant image (your puzzle input). The image includes empty space (.) and galaxies (#). For example:

...#......
.......#..
#.........
..........
......#...
.#........
.........#
..........
.......#..
#...#.....

The researcher is trying to figure out the sum of the lengths of the shortest path between every pair of galaxies. However, there's a catch: the universe expanded in the time it took the light from those galaxies to reach the observatory.

Due to something involving gravitational effects, only some space expands. In fact, the result is that any rows or columns that contain no galaxies should all actually be twice as big.

In the above example, three columns and two rows contain no galaxies:

   v  v  v
 ...#......
 .......#..
 #.........
>..........<
 ......#...
 .#........
 .........#
>..........<
 .......#..
 #...#.....
   ^  ^  ^

These rows and columns need to be twice as big; the result of cosmic expansion therefore looks like this:

....#........
.........#...
#............
.............
.............
........#....
.#...........
............#
.............
.............
.........#...
#....#.......

Equipped with this expanded universe, the shortest path between every pair of galaxies can be found. It can help to assign every galaxy a unique number:

....1........
.........2...
3............
.............
.............
........4....
.5...........
............6
.............
.............
.........7...
8....9.......

In these 9 galaxies, there are 36 pairs. Only count each pair once; order within the pair doesn't matter. For each pair, find any shortest path between the two galaxies using only steps that move up, down, left, or right exactly one . or # at a time. (The shortest path between two galaxies is allowed to pass through another galaxy.)

For example, here is one of the shortest paths between galaxies 5 and 9:

....1........
.........2...
3............
.............
.............
........4....
.5...........
.##.........6
..##.........
...##........
....##...7...
8....9.......

This path has length 9 because it takes a minimum of nine steps to get from galaxy 5 to galaxy 9 (the eight locations marked # plus the step onto galaxy 9 itself). Here are some other example shortest path lengths:

    Between galaxy 1 and galaxy 7: 15
    Between galaxy 3 and galaxy 6: 17
    Between galaxy 8 and galaxy 9: 5

In this example, after expanding the universe, the sum of the shortest path between all 36 pairs of galaxies is 374.

Expand the universe, then find the length of the shortest path between every pair of galaxies. What is the sum of these lengths?


--- Part Two ---

The galaxies are much older (and thus much farther apart) than the researcher initially estimated.

Now, instead of the expansion you did before, make each empty row or column one million times larger. That is, each empty row should be replaced with 1000000 empty rows, and each empty column should be replaced with 1000000 empty columns.

(In the example above, if each empty row or column were merely 10 times larger, the sum of the shortest paths between every pair of galaxies would be 1030. If each empty row or column were merely 100 times larger, the sum of the shortest paths between every pair of galaxies would be 8410. However, your universe will need to expand far beyond these values.)

Starting with the same initial image, expand the universe according to these new rules, then find the length of the shortest path between every pair of galaxies. What is the sum of these lengths?


"""

# LIBRARIES
import time
import pathlib
import os
from math import *
from functools import *
from sys import *
from collections import *
from queue import *

# MODULES
from _test import test

# OPTIONAL VARIABLES
DISPLAY_EXTRA_INFO = True
# DISPLAY_EXTRA_INFO = False

def manhattan_distance_with_expanding_empty_space(part, input_str, DEBUG = False, factor = 2):

  # CONSTANTS

  GALAXY, SPACE = '#', '.'


  # UTILITY

  def manhattan_distance(r1, c1, r2, c2):
    return abs(r2 - r1) + abs(c2 - c1)


  # PARSE INPUT DATA

  input_arr = input_str.split('\n')
  H = len(input_arr)
  W = len(input_arr[0])


  # DATA STRUCTURES

  # Create map
  MAP = [ [c for c in row] for row in input_arr ]
  
  # Create and populate collection of galaxies
  GALAXIES = []
  for r in range(H):
    for c in range(W):
      if MAP[r][c] == GALAXY:
        GALAXIES.append((r, c))

  # Create and populate collection of empty rows and cols
  EMPTY_ROWS = []
  EMPTY_COLS = []
  for r in range(H):
    if all([x == SPACE for x in MAP[r]]):
      EMPTY_ROWS.append(r)
  for c in range(W):
    if all([MAP[r][c] == SPACE for r in range(H)]):
      EMPTY_COLS.append(c)


  # ANALYZE

  total = 0

  for i in range(len(GALAXIES) - 1):
    for j in range(i + 1, len(GALAXIES)):

      # Add basic manhattan distance
      r1, c1 = GALAXIES[i]
      r2, c2 = GALAXIES[j]
      total += manhattan_distance(r1, c1, r2, c2)

      # Account for expansion of empty rows and cols
      for r in EMPTY_ROWS:
        if r1 < r < r2:                                   # NOTE: r1 < r2, since galaxies were discovered in row order
          total += (factor - 1)                           # don't forget to subtract 1!
      for c in EMPTY_COLS:
        if min(c1, c2) < c < max(c1, c2):                 # NOTE: c1 may not be less than c2
          total += (factor - 1)                           # don't forget to subtract 1!

  return total


# TEST CASES

test_num = [1]
test_input = None
test_expected = None
func = manhattan_distance_with_expanding_empty_space
skipped_tests = set([ 2, 3, 4, 5 ])
skipped_tests = set([ 3, 4, 5 ])
skipped_tests = set([ 5 ])
skipped_tests = set([  ])
lowest_test = 0
highest_test = 0

FILENAME = pathlib.Path(__file__).name
FILENAME_SPLIT = FILENAME.split('.py')[0].split('day')
DAY_NUM = FILENAME_SPLIT[1] if len(FILENAME_SPLIT) > 1 else None
CURRENT_DIR = os.path.dirname(os.path.realpath(__file__))
INPUT_PATH = CURRENT_DIR + '\\' + (f"day{DAY_NUM}-input.txt" if DAY_NUM else 'template-input.txt')
actual_input = open(INPUT_PATH, 'r').read()

sample_input = """...#......
.......#..
#.........
..........
......#...
.#........
.........#
..........
.......#..
#...#....."""

# Test case 1
test_input = {
  'part': 1,
  'input_str': sample_input,
  'DEBUG': True,
}
test_expected = 374
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)

# Test case 2
test_input = {
  'part': 1,
  'input_str': actual_input,
}
test_expected = 9723824
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)

# Test case 3
test_input = {
  'part': 2,
  'input_str': sample_input,
  'DEBUG': True,
  'factor': 10,
}
test_expected = 1030
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)

# Test case 4
test_input = {
  'part': 2,
  'input_str': sample_input,
  'DEBUG': True,
  'factor': 100,
}
test_expected = 8410
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)

# Test case 5
test_input = {
  'part': 2,
  'input_str': actual_input,
  'DEBUG': False,
  'factor': 1000000,
}
test_expected = 731244261352
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)